[francisyzy] iP

francisyzy opened the pull request:
# Duke 

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Are you tired of feeling overwhelmed by your daily tasks, deadlines, and events? Introducing Duke, the ultimate Todo, Deadline, and Event manager!

With Duke, you can easily organize your schedule and stay on top of your to-do list. Create tasks and set deadlines, and never miss an important event again.

Duke's user-friendly interface makes it easy to keep track of your progress and prioritize your tasks. Plus, with its advanced reminder system, you'll never forget a deadline or event again.

Duke also allows you to collaborate with your team members, so you can work together to achieve your goals.

Stop feeling stressed and start feeling in control with Duke. Try it now and experience the difference for yourself!

All you need to do is,
1. download it from [here](https://github.com/francisyzy/ip/releases/tag/A-Jar).
1. run `java -jar Duke.jar` in the directory where *Duke.jar* is downloaded
1. add your tasks.
1. let it manage your tasks for you :wink:

And it is **FREE**!

Features:
- [x] Managing todos
- [x] Managing deadlines
- [x] Managing events
- [x] Searching tasks
- [ ] Reminders (coming soon)
- [ ] GUI (coming soon)

---

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java{.no-line-numbers}
public class Main {
    public static void main(String[] args) {
        new Duke("data/tasks.txt").run();
    }
}
```

Commit:
Add Gradle support

francisyzy committed:
Update gitignore

francisyzy committed:
Add bindings to run current java file

francisyzy committed:
Add echo

francisyzy committed:
Level-1

francisyzy committed:
Level-2

francisyzy committed:
Level-3

francisyzy committed:
Level-4

francisyzy committed:
Catch more errors

francisyzy committed:
Add custom exception handling

francisyzy committed:
Support deleting from the list

francisyzy committed:
Trim strings

Remove whitespace before and after the user inputs

francisyzy committed:
Add tests

francisyzy committed:
Seperate it into different files

francisyzy committed:
Update test case

francisyzy committed:
Add enums

francisyzy committed:
Move Strings to enum

francisyzy committed:
Update tests

francisyzy committed:
Move check empty title\n\nMore OOP Style\n\nDont stack everything at the main duke file

francisyzy committed:
Add clear command

Clears the task list. Suitable for running consistant tests

francisyzy committed:
Add new constructor for already done task

So that when you load existing task list, the tasks would've already been completed

francisyzy committed:
Add automatic duke_data save

So that the data will not only be in memory but also a txt file.

Currently it is at current working directory, the files will be saved
at different places when running in different enviroments.

Also ignored the duke_data.txt in gitignore, user data not stored in VC

TODO: Figure a better place to store the file.

francisyzy committed:
Add datetime awareness

TODO: more checks to ensure data is correct
- [ ] end cannot be earlier than start
- [ ] throw error messages to user when parse fail

francisyzy committed:
Update test case and error message

Catches and throws error back to user when these two fault are found:
    - [X] end cannot be earlier than start
    - [X] throw error messages to user when parse fail

francisyzy committed:
Merge pull request #2 from francisyzy/branch-Level-7

Branch level 7

francisyzy committed:
Merge pull request #3 from francisyzy/branch-Level-8

Branch level 8

francisyzy committed:
Update test case and other branch

francisyzy committed:
Add exception conversion

Catch LocalDateTime parse exception and repackage it as DukeException

francisyzy committed:
Update testCase

Reset the list at the start

Incase of a failed run previously.

francisyzy committed:
Refactor Duke.java

Spilt it out into different java files

Make it more OOP

francisyzy committed:
Fix #5 Bug where [ ] is not cleared

francisyzy committed:
Remove wildcard imports

Close #4

francisyzy committed:
Move classes into packages

francisyzy committed:
Merge branch 'add-gradle-support'

francisyzy committed:
Fix boolean var name

francisyzy committed:
Fix runtest scripts

After packaging it, directory problems

Also forgot to import correctly in one java file

francisyzy committed:
Fix IndexOutOfBoundsException

francisyzy committed:
Fix level-8 print format

Also update test cases

francisyzy committed:
Fix time formatting

francisyzy committed:
Add JUnit Tests

Minimal goal. Will do scretch after finish other tasks

francisyzy committed:
Add Find

francisyzy committed:
Fix import ordering

Fix NewlineAtEndOfFile

francisyzy committed:
Add java docs

francisyzy committed:
Add CheckStyle

It is done using gradle

Cannot get the vscode extension to work but ymmv

francisyzy committed:
Fix merge issue

Add JavaDocs for Level9

francisyzy committed:
Add better JavaDocs

Clean up code duplication

Fix text ui test changes

francisyzy committed:
Init javafx

Nicklelodeon commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,50 @@
+    /**
+     * Runs the chatbot
+     */
+    public void run() {
+        ui.showWelcome();
+        boolean isExit = false;
+        while (!isExit) {
+            try {
+                String fullCommand = ui.readCommand();
+                Command c = Parser.parse(fullCommand);
+                c.execute(tasks, ui, storage);
+                isExit = c.isExit();
+            } catch (DukeException e) {
+                ui.showError(e.getMessage());
+            }
+        }
+    }
```
I like how you abstracted the different tasks to each class so that the code looks neat!

Nicklelodeon commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,97 @@
+                    throw new DukeException(Views.MISSING_ARGS_ERR_STRING.eng());
+                }
+                String title = input.substring(Commands.EVENT.cmd().length(), input.indexOf(Commands.FROM.cmd()));
+                String from = input.substring(input.indexOf(Commands.FROM.cmd()), input.indexOf(Commands.TO.cmd()));
```
Some of the lines (line 56, 63, 64) have a line length more than the soft limit of 110 characters. Although it is still less than the hard limit of 120 characters, perhaps the lines could be separated with a line break to improve readability?

Nicklelodeon commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,102 @@
+ * TaskList object to store the list of tasks
+ */
+public class TaskList {
+    private ArrayList<Task> tasksList;
```
I like how you used the plural form (tasks) when representing a collection of objects as mentioned in the coding guidelines, as I believe omitting plural form is a common mistake when naming variables.

Nicklelodeon commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,102 @@
+     *
+     * @param taskNo int index of task in the ArrayList
+     */
+    public void remove(int taskNo) {
```
As taskNo might be informal english and could be confused with the word "No", perhaps the variable name taskNo could be changed to taskNumber?

francisyzy replied:
Good catch! I'll use taskNum for less keystrokes

Nicklelodeon commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,46 @@
+ * Command: Deletes the Task given
+ */
+public class DeleteCommand extends Command {
+    private int taskNo;
```
As taskNo might be informal english and could be confused with the word "No", perhaps the variable name taskNo could be changed to taskNumber?

Nicklelodeon commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,46 @@
+     * @throws DukeException
+     */
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws DukeException {
+        ArrayList<Task> result = tasks.search(query);
```
Should the variable name result be changed to results as it contains a collection of Tasks?

Nicklelodeon commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,85 @@
+    /**
+     * Mark the task as not done
+     */
+    public void unmarkAsDone() {
```
Perhaps the method name could be changed to markAsUndone()?

Nicklelodeon reviewed (commented):
Overall, I found your code clean and easy to read! Most of the coding standards were adhered to too! 👍
Some cosmetic changes (regarding line length and variable names) were suggested.

francisyzy committed:
Fix minor style issues

Suggestions from:
https://github.com/nus-cs2103-AY2223S2/ip/pull/35#pullrequestreview-1274457225

francisyzy committed:
Add MVP Echo GUI

francisyzy committed:
Add Duke functions inside GUI

TODO:
- [ ] Fix string formatting
- [ ] Fix too long of a string

francisyzy committed:
Add img credits

francisyzy committed:
Fix changes not rebased properly

francisyzy committed:
Add minor styling change

francisyzy committed:
Add assertions

Close #1

francisyzy committed:
Fix arrowhead style code

francisyzy committed:
Merge pull request #6 from francisyzy/branch-A-Assertions

Add assertions

francisyzy committed:
Merge branch 'master' of github.com:francisyzy/ip into branch-A-CodeQuality

francisyzy committed:
Merge pull request #7 from francisyzy/branch-A-CodeQuality

Fix arrowhead style code

francisyzy committed:
Add Varargs

francisyzy committed:
Fix search by Add Streams

When introducing varargs to search for tasks, two bugs were created:
1. Broken search results due to empty string that match .contains()
2. Duplicated results

To fix 1. streams is used to remove empty string from varargs

To fix 2. HashSet to ensure only unique item are added, converted to
ArrayList afterwards to maintain compatibility with method return type

Solution was implemented with the help of Stackoverflow as Streams hard

francisyzy committed:
Add more assertions at Ui.java

totally missed out on this

francisyzy committed:
Improve CodeQuality

Remove unnecessary exception throws

Fix one error message not being abstracted properly, Duke.java directly
asking UI to print that error when it should've have been passed into it

Add JavaDocs, somehow missed out one when making new method

Reduce code duplication

Fix arrowhead style @ 8a9cc9481e085205909851efa5befeef31559ee7

francisyzy committed:
Merge pull request #14 from francisyzy/branch-A-Assertions

Add more assertions at Ui.java

francisyzy committed:
Merge branch 'master' into branch-A-CodeQuality

francisyzy committed:
Merge pull request #13 from francisyzy/branch-A-CodeQuality

Improve CodeQuality

francisyzy committed:
Merge pull request #12 from francisyzy/branch-A-Streams

Fix search by Add Streams

francisyzy committed:
Add CI

francisyzy committed:
Fix text-ui-test expected case and CI

francisyzy committed:
Fix runtest.bat

francisyzy committed:
Fix windows junit test

Issue caused by DOS Carriage return standard vs unix.
One uses \r\n whereas the other just use \n

francisyzy committed:
Add MassOps

francisyzy committed:
Add tests and fix bugs

Reduce code duplication

francisyzy committed:
Add MassOps

Merge pull request #15 from francisyzy/branch-C-MassOps

Current: Lack of MassOps feature

Why: Inefficient task management

Solution: Introduce MassOps

Reason: Improve productivity and efficiency

Info: MassOps allows deletion, marking, and unmarking of multiple tasks. 
Test added to test for proper function.
Go to following link for problem source:
https://nus-cs2103-ay2223s2.github.io/website/se-book-adapted/projectDuke/index.html#c-massops

francisyzy committed:
Add more JUnit tests

francisyzy committed:
Fix UI response for searching and list.

Replies the right message. Update tests accordingly.

Add test for search.

francisyzy committed:
Add better gui

Changes:
- Improve padding
- Change font type for users and duke replies
- Add error font diff
- Add graceful exit

See also: https://nus-cs2103-ay2223s2.github.io/website/admin/ip-w6.html#a-bettergui

francisyzy committed:
Add error handling for mark/unmark

Will notify the user to inform them that they have already marked the
task if it was already marked. Previously it will just tell the user
that it is marked. Now will error out.

See: https://nus-cs2103-ay2223s2.github.io/website/admin/ip-w6.html#a-moreerrorhandling

francisyzy committed:
Add joke languages

User can choose between the following language:
lolcat
pirate
shakespeare

They can change language using this command `set language` for example
`set lolcat` to use lolcat language.

francisyzy committed:
Add graceful exit if CLI cannot detect nextline

Ctr + D will send empty line which crashes the Duke. Catches the problem
and send exit message and System.Exit gracefully

francisyzy committed:
Fix out of bound error

Also changed some methods to static. Should use more static methods...

francisyzy committed:
Add help message for user

Also change test to directly pull from enum so that strings only need to
change one side of the text without breaking all the tests.

See also https://nus-cs2103-ay2223s2.github.io/website/se-book-adapted/projectDuke/index.html#c-help

francisyzy committed:
Add implements Comparable

So that the result output from searching will remain consistent when
testing.

francisyzy committed:
Remove assertError catch

Add more DukeException in places assertion was used.

Add DukeException to take in Views directly. Remove complexity of
Views.ENUM.str() when you can just pass in VIEWS

Reduce complexity at parser by abstracting it elsewhere. Trying to SLAP!

See this thread for more details: https://github.com/nus-cs2103-AY2223S2/forum/issues/152#issuecomment-1424107028

francisyzy committed:
Refactor Ui.java

Change most methods to static because they do not need to be passed one

Removed single use methods that is only used once, add method to
consolidate it wherever convenient

francisyzy committed:
Fix code quality

Fix #17

Implemented more SLAP

Fix Header Comments

francisyzy committed:
Add export to Markdown

francisyzy committed:
Update publish script

francisyzy committed:
Add user guide
